R/risk_filter.R

```diff
@@ -1,14 +1,80 @@
+#' to [`utils::available.packages()`]'s `filter` argument, or for use with
+#' the `available_packages_filter` option.
+#'
+#' @param ... Conditions to use when specifying a risk criteria. Uses
+#'   non-standard evaluation, treating risk field names as columns of package
+#'   metadata.
+#' @param fields A character vector of field names used in the provided `...`
+#'   conditions. When `NULL`, the default, all known risk field names will be
+#'   considered.
+#'
 #' @export
-risk_filter <- function(...) {
+risk_filter <- function(..., fields = NULL) {
   conditions <- rlang::quos(...)
 
+  if (is.null(fields)) {
+    all_fields_files <- list.files(
+      system.file(package = packageName()),
+      pattern = "RISKFIELDS",
+      recursive = TRUE,
+      full.names = TRUE
+    )
+
+    fields <- unique(unlist(lapply(dirname(all_fields_files), risk_fields)))
+  }
+
+  req_fields <- intersect(
+    unlist(lapply(conditions, all.names)),
+    fields
+  )
+
   # List of a function as needed by `available.packages()` for filtering.
-  list(function(pkgs) {
-    pkgs |>
-      tibble::as_tibble() |>
-      # TODO: Note that `pkgs` is the matrix returned by `available.packages()`
-      #       which is of type character. We will need to coerce the risk
-      #       metrics' columns to appropriate type.
-      dplyr::filter(!!!conditions)
-  })
+  list(
+    add = TRUE,
+    risk_filter = function(pkgs) {
+      # we'll use the parent call to `available.packages()` to re-evalauate
+      # available.packages as though the appropriate risk fields were included
+
+      # deduce which filters were used when calling available.packages
+      ap_call <- match.call(sys.function(-1), sys.call(-1), expand.dots = TRUE)
+      ap_filters <- if (!is.null(ap_call$filters)) {
+        eval(ap_call$filters, envir = parent.frame(2))
+      } else {
+        global_filters()
+      }
+
+      # find any filters that would have been applied up to this point
+      #   TODO: do we need to handle cases where risk_filter() result is
+      #   combined with other filters
+      #   (ie character values, multiple "risk_filters" keys?)
+      is_risk_filter <- function(name) name == "risk_filter"
+      idx_self_filter <- Position(is_risk_filter, names(ap_filters), 1L)
+      prior_filters <- utils::head(ap_filters, idx_self_filter - 1L)
+
+      # re-calculate our available.packages with required risk fields
+      ap_call$filters <- prior_filters
+      ap_call$fields <- eval(ap_call$fields, envir = parent.frame())
+      ap_call$fields <- unique(c(ap_call$fields, req_fields))
+      prior_pkgs <- eval(ap_call, envir = parent.frame(2))
+
+      # TODO: Note that `prior_pkgs` is the matrix returned by
+      # `available.packages()` which is of type character. We will need to
+      # coerce the risk metrics' columns to appropriate type.
+
+      prior_pkgs |>
+        tibble::as_tibble() |>
+        dplyr::filter(!!!conditions)
+    }
```

Comment:
Hi @dgkf,

I'm sorry to be replying only now!

- Given the implementation requirements, specifically the canonical use of `available.packages` and the `filters` parameter, I think this proposal looks good! :)
- I agree with your TODO notes. Perhaps we could raise a specific issue for each one and merge this PR. (@llrs, would like to chime in?)

On a higher-level note, I wonder if it might be beneficial to rethink the entire process of handling a curated catalog of packages from scratch, avoiding some of these workarounds. (I'm sorry to be bringing this up again :) ) This might involve moving away from the `available.packages` interface and decoupling the metadata fields about packages from the PACKAGES file. As we discussed earlier, if we had a way to refer to a package unambiguously (using a combination of package name, version, R version, repo, and arch), we could potentially create an "officially" verified set of packages---a catalogue,  materialised as a special S7 R object---that could be published regularly by third parties. Users could then select and install packages from these set(s). A sys admin of a pkg repo could then opt for providing the whole set of a verified catalogue, or a mixture of different catalogues from different sources. What do you think about keeping package repositories and catalogs separate? I would also distinguish verification and validation catalogues. Verification in this context would simply mean assessed packages on a range of criteria (as riskscore does it). This could vary significantly from discipline to discipline. Validation would be a subsequent step that decides which packages should be selected to usage.

Reply:
@ramiromagno handling a curated catalog of packages from scratch is easily done with tools like r-universe, or just providing a sanitized internal mirror. I am not sure I understood (yet) who would be running/maintaining this verified repository, but here this is already decoupled: all packages in the future pharmaR repo(s) are assessed, but only the filters selected are validated for a given purpose (depending on the filtering). 

I see some benefits of having a separated record if the same external package is assessed with multiple tools and versions of tools that provide different outcomes. Then it could make sense but it adds more complexity and variability and with little gain (keeping records with riskscore_v0.9.0 and riskscore_v50.0.0 for example)

Reply:
While drafting an email to r-devel I was thinking more about pulling data separately, and I don't see it as bad as I initially thought of it. If the data could be downloaded at filtering steps we could avoid the problem with the fields at the `available.packages` step completely: 
So something like 
`filters = filter_risk(function(x, y = "https://company.CRAN.evals.data") {merge(x, readRDS(y)); ...))`.
I am not sure if there is a way to pass arguments to these functions or they could just read from an option or environmental variable ("pharmakg_qc_data" or similar) set by the R admin.
This would require to download this data at each install, but it might be safer than just calling eval on anything that goes through filters.

Reply:
> data could be downloaded at filtering steps we could avoid the problem with the fields at the available.packages step completely

I think we should split this off into a separate issue discussion.

Just briefly, I definitely agree that from a software this is by far the preferred solution. However, there are some concerns that have surfaced repeatedly from companies in the regulatory space regarding the impact of the system where the quality is measured. By coupling our metrics to the management of a repository, we are able to couple metrics to the image that was used for evaluating packages. Maybe we could use an image digest as one of the keys for merging to get around this?

I think it's a very interesting idea, but there are a ton of considerations to lock down.